In [8]:
from numpy import mean,std
from matplotlib import pyplot as plt
from sklearn.model_selection import KFold
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D,MaxPooling2D,Dense,Flatten
from tensorflow.keras.optimizers import Adam

In [10]:
# load dataset
(trainX, trainy), (testX, testy) = mnist.load_data()
# summarize loaded dataset
print('Train: X=%s, y=%s' % (trainX.shape, trainy.shape))
print('Test: X=%s, y=%s' % (testX.shape, testy.shape))


Train: X=(60000, 28, 28), y=(60000,)
Test: X=(10000, 28, 28), y=(10000,)


In [11]:
(trainX, trainY), (testX, testY) = mnist.load_data()

# reshape dataset to have a single channel
trainX = trainX.reshape((trainX.shape[0], 28, 28, 1))
testX = testX.reshape((testX.shape[0], 28, 28, 1))


In [12]:
from tensorflow.keras.utils import to_categorical

# one hot encode target values
trainY = to_categorical(trainY)
testY = to_categorical(testY)

In [13]:
# convert from integers to floats
train_norm = trainX.astype('float32')
test_norm = testX.astype('float32')
# normalize to range 0-1
trainX = train_norm / 255.0
testX = test_norm / 255.0

**Model 1**

In [46]:
def define_model_1():
	model = Sequential()
	model.add(Flatten(input_shape=(28, 28)))
	model.add(Dense(200, activation='relu'))
	model.add(Dense(10, activation='softmax'))
	# compile model
	opt = Adam(learning_rate=0.01)
	model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
	return model

In [47]:
# evaluate a model using k-fold cross-validation
def evaluate_model_1():
    scores, histories = list(), list()
    # define model
    model = define_model_1()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories


In [48]:
scores, histories = evaluate_model_1()

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_10 (Flatten)        (None, 784)               0         
                                                                 
 dense_30 (Dense)            (None, 200)               157000    
                                                                 
 dense_31 (Dense)            (None, 10)                2010      
                                                                 
Total params: 159,010
Trainable params: 159,010
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2404 - accuracy: 0.9291 - val_loss: 0.1592 - val_accuracy: 0.9543
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1699 - accuracy: 0.9532 - val_loss: 0.1655 - val_accuracy: 0.9546
Epoch 3/10
1875/18

**Model 2**

In [52]:
def define_model_2():
    model = Sequential()
    model.add(Flatten(input_shape=(28, 28)))
    model.add(Dense(190, activation='relu'))
    model.add(Dense(25, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(10, activation='softmax'))
    # compile model
    opt = Adam(learning_rate=0.01)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model


In [53]:
# evaluate a model using k-fold cross-validation
def evaluate_model_2():
    scores, histories = list(), list()
    # define model
    model = define_model_2()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

Execute the Second Model Training

In [54]:
scores, histories = evaluate_model_2()

Model: "sequential_12"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_12 (Flatten)        (None, 784)               0         
                                                                 
 dense_36 (Dense)            (None, 190)               149150    
                                                                 
 dense_37 (Dense)            (None, 25)                4775      
                                                                 
 dense_38 (Dense)            (None, 60)                1560      
                                                                 
 dense_39 (Dense)            (None, 10)                610       
                                                                 
Total params: 156,095
Trainable params: 156,095
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [====================


**Report**

---



Two Deep Learning Models with softmax layer activation at the top and 'relu' on other layers using Neural Network architecture was created on the MNIST dataset.

*MODEL 1*: \\
Number of NN layers = 2 \\
Validation Accuracy = 96.730 \\
Validation Loss = 0.2315 \\
Number of Trainable Parameters = 159,010

*MODEL 2*: \\
Number of NN layers = 4 \\
Validation Accuracy = 97.110 \\
Validation Loss = 0.1353 \\
Number of Trainable Parameters = 156,095

*Model 2* has better validation accuracy than Model 1 hence Model 2 has better performance.
